File: DS/Trees/Binary tree/CPP/level_order_traversal.cpp
#include <iostream>
#include <queue>
using namespace std;

// node class

class btNode
{
public:
    int data;
    btNode* left;
    btNode* right;

    // Constructor to create a new node
    btNode(int num){
        data = num;
        left = right = NULL;
    }
};

// insert function

btNode* Insert(btNode* pointer, int data){
    if(pointer == NULL){
        pointer = new btNode(data);
    }

    else if(data<=pointer->data){
        pointer->left = Insert(pointer->left, data);
    }

    else{
        pointer->right = Insert(pointer->right, data);
    }

    return pointer;
}

// levelorder traversal

void levelOrder(btNode* root){
    if(root == NULL){
        cout<<"Tree is empty";
        return;
    }

    queue<btNode*> Q;
    Q.push(root);

    while (!Q.empty())
    {
        btNode* node = Q.front();
        cout<<node->data<<" ";

        if(node->left!=NULL){Q.push(node->left);}
        if(node->right!=NULL){Q.push(node->right);}
        Q.pop();
    }
}

//main

int main(){
    btNode* root = NULL;
    root = Insert(root, 10);
    root = Insert(root, 25);
    root = Insert(root, 8);
    root = Insert(root, 9);
    root = Insert(root, 7);
    root = Insert(root, 15);
    root = Insert(root, 30);

    levelOrder(root);
}

// output: 10 8 25 7 9 15 30

File: DS/Trees/Binary tree/PYTHON/level_order_traversal.py
# node class
class BtNode:
    def __init__(self, num):
        self.data = num
        self.left = None
        self.right = None


# insert function

def Insert(pointer, data):
    if pointer is None:
        pointer = BtNode(data)

    elif data <= pointer.data:
        pointer.left = Insert(pointer.left, data)

    else:
        pointer.right = Insert(pointer.right, data)

    return pointer


# levelorder traversal
def levelorder(root):
    Q = []
    Q.append(root)

    while len(Q) != 0:
        node = Q.pop(0)

        print(node.data, end=" ")

        if node.left is not None:
            Q.append(node.left)

        if node.right is not None:
            Q.append(node.right)


if __name__ == "__main__":
    root = None

    root = Insert(root, 10)
    root = Insert(root, 8)
    root = Insert(root, 9)
    root = Insert(root, 7)
    root = Insert(root, 25)
    root = Insert(root, 30)
    root = Insert(root, 15)

    print("Levelorder:")
    levelorder(root)

# output:
# 10 8 25 7 9 15 30 
